### PR TITLE
Conductor: implementation of several test cases

### LINKED ISSUE #704: Conductor: Verify not existent block, this should work fine
//Verify not existent block, this should work fine, since we only formally verify verification_tickets
```
(T0) Leader_0: send Verification0_X
(T0 + δ) Replica_0: verify and forget

```

### LINKED ISSUE #707: Conductor: Half of nodes are down, generator proposes but can't get enough verification tickets
//Half of nodes are down, generator proposes but can't get enough verification tickets, restarts. Soon half of nodes get up and restarts round.
```
(T0) Replica_i: down, 0<=i<1/2f
(T0) Leader0_0: send Proposal0_0
(T0 + δ + Δ) Replica_0: send VerificationTicket0_0
(T0 + timeout) Replica_j: send VRFShare(timeout), 1/2f<j<=f
(T0 + 4*timeout) Replica_i: get up, send VRFShare(timeout), 0<=i<1/2f
(T0 + 4*timeout + δ): Leader0_1: send Proposal0_1
```

### LINKED ISSUE #708: Conductor: Malicious replicas start to send VRFShares for current round as if they get up after restart
//Malicious replicas start to send VRFShares for the current round as if they get up after restart
```
(T0) Leader_0: send Proposal0_0
(T0 + δ) Replica_i(Adv): send VRFShare(timeout), 0<=i<1/3f
(T0 + δ + Δ) Replica_j: send VerificationTicket0_j 
```

### LINKED ISSUE #709: Conductor: Miner sends previously proposed block used
//Miner sends previously proposed block used in the notarized round with one/several blocks notarized in it to propose in a new round
```
(T0) Leader_0: send Proposal0_0, Proposal0_1
(T0 + δ + Δ) Replica_j: send VerificationTicket0_j, VerificationTicket1_j  
(T0 + 2δ + Δ) Replica_j: send Notarization0_0, Notarization0_1 
(T0 + 3δ + Δ) Leader_1: send Proposal0_0
(T0 + 4δ + Δ) Replica_j: reject Proposal0_0
```

### LINKED ISSUE #710: Conductor: Miner sends previous notarization used in notarized/finalized round in new round
//Miner sends previous notarization used in notarized/finalized round in new round
```
(T0) Replica_j: obtain Notarization0_0
(T1) Leader_1: send Proposal1_0
(T1 + δ + Δ) Replica_j: send VerificationTicket1_j  
(T1 + 2δ + Δ) Replica_j: send Notarization1_1 
(T1 + 3δ + Δ) Replica_j: reject Notarization0_0

```

### LINKED ISSUE #817: Conductor: Notarize not existent block and fetch it
```
// Notarize not existent block and fetch it
(T0) Replica_i: send Verification_i_X  0<i<=2f/3 + 1
(T0 + δ) Replica_1: fetch Block_i
(T0 + 2*δ) Replica_1: send Notarization_i
```

### PR DESCRIPTION
**Pull request by MurashovVen**
Test cases:

1. Closes #704 
```
// Verify not existent block, this should work fine, since we only formally verify verification_tickets
(T0) Replica_0: send Verification0_X
(T0 + δ) Replica_1: check and forget
```

2. Closes #817 
```
// Notarize not existent block and fetch it
(T0) Replica_i: send Verification_i_X  0<i<=2f/3 + 1
(T0 + δ) Replica_1: fetch Block_i
(T0 + 2*δ) Replica_1: send Notarization_i
```

3. Closes #709
```
// Miner sends previous proposed block used in notarized round with one/several blocks notarized in it to propose in new round
(T0) Leader_0: send Proposal0_0 (after getting first Notarisation)
(T0) Leader_1: send Proposal0_1
(T0 + δ + Δ) Replica_j: send VerificationTicket1_j
(T0 + 2δ + Δ) Replica_j: send Notarization0_1
(T0 + 3δ + Δ) Leader_1: resend Proposal0_0 (after getting Proposal0_0)
(T0 + 4δ + Δ) Replica_j: reject Proposal0_0
```

4. Closes #710
```
// Miner sends previous notarization used in notarized/finalized round in new round
(T0) Replica_j: obtain Notarization0_0
(T1) Leader_1: send Proposal1_0
(T1 + δ + Δ) Replica_j: send VerificationTicket1_j
(T1 + 2δ + Δ) Replica_j: send Notarization1_1
(T1 + 3δ + Δ Replica_j: send Notarisation0_0
(T1 + 4δ + Δ) Replica_j+1: reject Notarization0_0
```

5. Closes #708
```
// Malicious replicas start to send VRFShares for current round as if they get up after restart
(T0) Leader_0(Adv): send VRFShare(timeout), 0<=i<1/3f
```

6. Closes #707
```
// Half of nodes are down, generator proposes but can't get enough verification tickets, restarts. Soon half of nodes get up and restarts round.
(T0) Replica_i: down, 0<=i<1/2f
(T0) Leader0_0: send Proposal0_0
(T0 + δ + Δ) Replica_0: send VerificationTicket0_0
(T0 + timeout) Replica_j: send VRFShare(timeout), 1/2f<j<=f
(T0 + 4*timeout) Replica_i: get up, send VRFShare(timeout), 0<=i<1/2f
(T0 + 4*timeout + δ): Leader0_1: send Proposal0_1
```

### CODE REVIEW

File: code/go/0chain.net/chaincore/chain/handler.go [hunk 1]
```diff
@@ -36,37 +36,100 @@ import (
...
+	transactionEntityMetadata := datastore.GetEntityMetadata("txn")
+	m := map[string]func(http.ResponseWriter, *http.Request){
+		"/v1/chain/get": common.Recover(
+			common.ToJSONResponse(
+				memorystore.WithConnectionHandler(
+					GetChainHandler,
+				),
+			),
+		),
+		"/v1/chain/put": common.Recover(
```
Comment (dabasov) [inline code/go/0chain.net/chaincore/chain/handler.go:53]: Do you think one liner is more in go style?

File: code/go/0chain.net/conductor/cases/breaking_single_block.go [hunk 2]
```diff
@@ -31,22 +30,11 @@ type (
 		FirstSentBlockHash  string `json:"first_sent_block_hash"`
 		SecondSentBlockHash string `json:"second_sent_block_hash"`
 	}
-
-	BreakingSingleBlockResult struct {
```
Comment (dabasov) [inline code/go/0chain.net/conductor/cases/breaking_single_block.go:35]: why do we delete it?

File: code/go/0chain.net/conductor/cases/notarising_non_existent_block.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		return false, ctx.Err()
+
+	case <-prepared:
+		return n.check()
+	}
+}
+
+func (n *NotarisingNonExistentBlock) check() (success bool, err error) {
+	notBlocks := n.res.NotarisedBlocks
+	if len(notBlocks) != 1 || notBlocks[0].Rank != 0 {
```
Comment (dabasov) [inline code/go/0chain.net/conductor/cases/notarising_non_existent_block.go:62]: how do we control, that second block from lower ranked miner is not notarized before top ranked?

File: code/go/0chain.net/conductor/cases/resend_notarisation.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	case <-ctx.Done():
+		return false, errors.New("cases state is not prepared, context is done")
+
+	case <-prepared:
+		return n.check()
+	}
+}
+
+func (n *ResendNotarisation) check() (success bool, err error) {
+	if !n.result.IsFinalised {
```
Comment (dabasov) [inline code/go/0chain.net/conductor/cases/resend_notarisation.go:59]: we should be aware that round can be finalized not only directly but also via finalized sibling. so if Rk-1 is finalized and Rk+1 is finalized too, then there is Rk round that is also finalized and Rk-1 -prev-> Rk -prev-> Rk+1

File: code/go/0chain.net/miner/protocol_receive_integration_tests.go [hunk 5]
```diff
@@ -83,3 +107,131 @@ func configureBreakingSingleBlock(firstBlockHash, secondBlockHash string) error
...
+	}
+	resendNotarisationCfg.Unlock()
+
+	mc.handleNotarizationMessage(ctx, msg)
+}
+
+func isObtainingNotarisation(round int64) bool {
+	cfg := crpc.Client().State().ResendNotarisation
+	nodeType, typeRank := getNodeTypeAndTypeRank(round)
+	return cfg != nil && round == cfg.OnRound-2 && !cfg.Resent &&
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_receive_integration_tests.go]: should write comment on this

File: code/go/0chain.net/miner/protocol_receive_integration_tests.go [hunk 6]
```diff
@@ -83,3 +107,131 @@ func configureBreakingSingleBlock(firstBlockHash, secondBlockHash string) error
...
+		((nodeType == generator) == cfg.ByGenerator) && cfg.ByNodeWithTypeRank == typeRank
+}
+
+func isResendingNotarisation(round int64) bool {
+	cfg := crpc.Client().State().ResendNotarisation
+	nodeType, typeRank := getNodeTypeAndTypeRank(round)
+	return cfg != nil && round == cfg.OnRound-1 &&
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_receive_integration_tests.go]: should write comment on this

File: code/go/0chain.net/miner/protocol_receive_integration_tests.go [hunk 7]
```diff
@@ -83,3 +107,131 @@ func configureBreakingSingleBlock(firstBlockHash, secondBlockHash string) error
...
+	if roundI == nil || !roundI.IsRanksComputed() {
+		return -1, -1
+	}
+
+	genNum := mc.GetGeneratorsNumOfRound(round)
+	isGenerator := mc.IsRoundGenerator(roundI, node.Self.Node)
+	nodeType, typeRank = generator, roundI.GetMinerRank(node.Self.Node)
+	if !isGenerator {
+		nodeType = replica
+		typeRank = typeRank - genNum
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_receive_integration_tests.go:247]: why?

File: code/go/0chain.net/miner/protocol_send_integration_tests.go [hunk 8]
```diff
@@ -98,6 +71,28 @@ func (mc *Chain) SendVRFShare(ctx context.Context, vrfs *round.VRFShare) {
 	}
 }
 
+func isSendingBadTimeoutVRFS(round int64) bool {
+	state := crpc.Client().State()
+	cfg := state.BadTimeoutVRFS
+	return cfg != nil && cfg.OnRound == round && state.IsMonitor
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_send_integration_tests.go:78]: comment please

File: code/go/0chain.net/miner/protocol_send_integration_tests.go [hunk 9]
```diff
@@ -200,26 +203,75 @@ func (mc *Chain) SendVerificationTicket(ctx context.Context, b *block.Block,
...
 }
 
-func (mc *Chain) sendDifferentBlocks(ctx context.Context, b *block.Block) {
+func isSendingDifferentBlocksFromFirstGenerator(r int64) bool {
+	mc := GetMinerChain()
+
+	currRound := mc.GetRound(r)
+	isFirstGenerator := currRound.GetMinerRank(node.Self.Node) == 0
+	testCfg := crpc.Client().State().SendDifferentBlocksFromFirstGenerator
+	return testCfg != nil && testCfg.OnRound == r && isFirstGenerator && currRound.GetTimeoutCount() == 0
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_send_integration_tests.go]: and this

File: code/go/0chain.net/miner/protocol_send_integration_tests.go [hunk 10]
```diff
@@ -200,26 +203,75 @@ func (mc *Chain) SendVerificationTicket(ctx context.Context, b *block.Block,
...
+}
+
+func isSendingDifferentBlocksFromAllGenerators(r int64) bool {
+	mc := GetMinerChain()
+
+	currRound := mc.GetRound(r)
+	isGenerator := mc.IsRoundGenerator(mc.GetRound(r), node.Self.Node)
+	testCfg := crpc.Client().State().SendDifferentBlocksFromAllGenerators
+	return testCfg != nil && testCfg.OnRound == r && isGenerator && currRound.GetTimeoutCount() == 0
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_send_integration_tests.go:272]: and this

File: code/go/0chain.net/miner/protocol_send_integration_tests.go [hunk 11]
```diff
@@ -316,14 +354,18 @@ func createDataTxn(data string) (*transaction.Transaction, error) {
...
 
-func (mc *Chain) isSendingInsufficientProposals(r int64) bool {
+func isSendingInsufficientProposals(r int64) bool {
+	mc := GetMinerChain()
+
 	currRound := mc.GetRound(r)
 	isFirstGenerator := currRound.GetMinerRank(node.Self.Node) == 0
 	testCfg := crpc.Client().State().SendInsufficientProposals
-	return testCfg != nil && testCfg.Round == r && isFirstGenerator && currRound.GetTimeoutCount() == 0
+	return testCfg != nil && testCfg.OnRound == r && isFirstGenerator && currRound.GetTimeoutCount() == 0
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_send_integration_tests.go]: and this

File: code/go/0chain.net/miner/protocol_send_integration_tests.go [hunk 12]
```diff
@@ -346,9 +388,8 @@ func (mc *Chain) sendInsufficientProposals(ctx context.Context, b *block.Block)
...
 
-func (mc *Chain) isConfiguringSendInsufficientProposals(r int64) bool {
-	currRound := mc.GetRound(r)
-	isSecondGenerator := currRound.GetMinerRank(node.Self.Node) == 1 && mc.IsRoundGenerator(currRound, node.Self.Node)
-	cfg := crpc.Client().State().SendInsufficientProposals
-	return cfg != nil && cfg.Round == r && isSecondGenerator && currRound.GetTimeoutCount() == 0
+func isDelayingBlock(round int64) bool {
+	cfg := crpc.Client().State().ResendProposedBlock
+	nodeType, typeRank := getNodeTypeAndTypeRank(round)
+	return cfg != nil && cfg.OnRound == round && nodeType == generator && typeRank == 0
```
Comment (dabasov) [inline code/go/0chain.net/miner/protocol_send_integration_tests.go:397]: and this

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/conductor/cases/notarising_non_existent_block.go (hunk 3 above)
Comment (MurashovVen) [reply code/go/0chain.net/conductor/cases/notarising_non_existent_block.go:62]: Added mechanism to ignoring blocks generation phase in Generator1 (bbd1159)

File: code/go/0chain.net/miner/protocol_send_integration_tests.go (hunk 12 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_send_integration_tests.go:397]: Commented

File: code/go/0chain.net/miner/protocol_receive_integration_tests.go (hunk 5 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_receive_integration_tests.go]: Commented

File: code/go/0chain.net/miner/protocol_receive_integration_tests.go (hunk 6 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_receive_integration_tests.go]: Commented

File: code/go/0chain.net/miner/protocol_send_integration_tests.go (hunk 8 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_send_integration_tests.go:78]: Commented

File: code/go/0chain.net/miner/protocol_send_integration_tests.go (hunk 9 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_send_integration_tests.go]: Commented

File: code/go/0chain.net/miner/protocol_send_integration_tests.go (hunk 10 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_send_integration_tests.go:272]: Commented

File: code/go/0chain.net/miner/protocol_send_integration_tests.go (hunk 11 above)
Comment (MurashovVen) [reply code/go/0chain.net/miner/protocol_send_integration_tests.go]: Commented

**Review by dabasov — APPROVED** [review]
lgtm

**Comment by MurashovVen** [comment]
closed by #990

**Pull request closed without merging**